File: ios/Hapi/Features/Chat/Blocks/ToolBodies.swift
import HapiProtocol
import HapiUI
import SwiftUI

/// Read-only inspector body: input rendering per tool kind + the result
/// section (the read-only slice of `web/src/components/ToolCard/views/`,
/// via the Android `ToolBodies` port):
///
/// - terminal family → command as a bash code block, stdout/stderr
///   terminal-styled;
/// - `Edit`/`MultiEdit` structured edits → before/after code blocks;
/// - `Write` → the written content as a code block;
/// - `CodexDiff` (and any input/result that parses as a unified diff) →
///   `DiffTextView`;
/// - `TodoWrite`/`update_plan` → checklist rows;
/// - Ask/RequestUserInput → questions + selected answers, read-only;
/// - anything else → pretty-printed JSON input, then the generic result.
struct ToolCallBody: View {
    let tool: ChatToolCall
    let basePath: String?
    @State private var sourceExpanded = false

    var body: some View {
        let questionTool = isQuestionDetailsTool(tool.name)
        let answers = questionTool ? tool.permission?.answers : nil
        VStack(alignment: .leading, spacing: 12) {
            if questionTool {
                QuestionToolBody(tool: tool)
            } else {
                SectionLabel(text: String(localized: "Input"))
                ToolInputSection(tool: tool, basePath: basePath)
                ToolResultSection(tool: tool)
            }
            if tool.input != nil || tool.result != nil || answers != nil {
                DisclosureGroup("Source", isExpanded: $sourceExpanded) {
                    if sourceExpanded {
                        if let input = tool.input {
                            SectionLabel(text: String(localized: "Input"))
                            GenericJSONInput(input: input)
                        }
                        if let result = tool.result {
                            SectionLabel(text: String(localized: "Result"))
                            GenericJSONInput(input: result)
                        }
                        if let answers {
                            SectionLabel(text: String(localized: "Answers"))
                            GenericJSONInput(input: answers)
                        }
                    }
                }
            }
        }
    }
}

private struct QuestionToolBody: View {
    let tool: ChatToolCall
    @State private var details: QuestionToolDetails?

    var body: some View {
        Group {
            if let details {
                SectionLabel(text: details.hasAnswers
                    ? String(localized: "Questions & Answers") : String(localized: "Input"))
                if details.questions.isEmpty {
                    GenericJSONInput(input: tool.input)
                } else {
                    QuestionDetailsView(questions: details.questions)
                }
                if details.showResult { ToolResultSection(tool: tool) }
            } else {
                ProgressView()
            }
        }
        .task(id: tool) {
            let tool = tool
            let next = await Task.detached(priority: .userInitiated) { questionToolDetails(tool) }.value
            guard !Task.isCancelled else { return }
            details = next
        }
    }
}

// MARK: - Input

private struct ToolInputSection: View {
    let tool: ChatToolCall
    let basePath: String?

    var body: some View {
        let input = tool.input
        let name = toolPresentationName(tool.name)
        if terminalToolNames.contains(name) {
            if let command = chatTerminalCommand(input) {
                ToolTextContent(language: "bash", code: command)
            } else {
                GenericJSONInput(input: input)
            }
        } else if name == "exec", let source = toolSourceInput(input, keys: ["code", "script"]) {
            ToolTextContent(language: "javascript", code: source)
        } else if name == "CodexPatch", let patch = toolSourceInput(input, keys: ["patch", "input", "command"]) {
            ToolTextContent(language: "diff", code: patch)
        } else if name == "Edit" {
            if let old = chatInputString(input, ["old_string"]),
               let new = chatInputString(input, ["new_string"]) {
                BeforeAfterView(
                    old: old,
                    new: new,
                    language: languageForPath(chatInputString(input, ["file_path", "path"]))
                )
            } else {
                GenericJSONInput(input: input)
            }
        } else if name == "MultiEdit" {
            multiEditBody(input)
        } else if name == "Write" {
            if let content = chatInputString(input, ["content", "text"]) {
                ToolTextContent(
                    language: languageForPath(chatInputString(input, ["file_path", "path"])),
                    code: content
                )
            } else {
                GenericJSONInput(input: input)
            }
        } else if name == "CodexDiff" {
            if let unified = chatInputString(input, ["unified_diff"]) {
                ToolDiffContent(text: unified)
            } else {
                GenericJSONInput(input: input)
            }
        } else if name == "TodoWrite" || name == "update_plan" {
            let items = checklistItems(input)
            if !items.isEmpty {
                VStack(alignment: .leading, spacing: 2) {
                    ForEach(Array(items.enumerated()), id: \.offset) { _, item in
                        Text("\(item.glyph) \(item.text)")
                            .font(.footnote)
                    }
                }
            } else {
                GenericJSONInput(input: input)
            }
        } else if name == "request_user_input_async" {
            QuestionsReadOnlyView(input: input)
        } else {
            GenericJSONInput(input: input)
        }
    }

    @ViewBuilder
    private func multiEditBody(_ input: JSONValue?) -> some View {
        let language = languageForPath(chatInputString(input, ["file_path", "path"]))
        if let edits = input?[chatKey: "edits"]?.chatArray, !edits.isEmpty {
            VStack(alignment: .leading, spacing: 8) {
                ForEach(Array(edits.enumerated()), id: \.offset) { index, edit in
                    if let old = chatInputString(edit, ["old_string"]),
                       let new = chatInputString(edit, ["new_string"]) {
                        VStack(alignment: .leading, spacing: 6) {
                            if edits.count > 1 {
                                SectionLabel(text: String(
                                    format: String(localized: "Edit %lld/%lld"),
                                    Int64(index + 1),
                                    Int64(edits.count)
                                ))
                            }
                            BeforeAfterView(old: old, new: new, language: language)
                        }
                    } else {
                        GenericJSONInput(input: edit)
                    }
                }
            }
        } else {
            GenericJSONInput(input: input)
        }
    }
}

private struct BeforeAfterView: View {
    let old: String
    let new: String
    let language: String?

    var body: some View {
        VStack(alignment: .leading, spacing: 6) {
            SectionLabel(text: String(localized: "Before"))
            ToolTextContent(language: language, code: old.isEmpty ? String(localized: "(empty)") : old)
            SectionLabel(text: String(localized: "After"))
            ToolTextContent(language: language, code: new.isEmpty ? String(localized: "(empty)") : new)
        }
    }
}

private struct GenericJSONInput: View {
    let input: JSONValue?

    var body: some View {
        switch input {
        case nil, .some(.null):
            EmptyView()
        case .some(.string(let text)):
            ToolTextContent(language: nil, code: text)
        case .some(let value):
            ToolJSONContent(value: value)
        }
    }
}

private struct ToolJSONContent: View {
    let value: JSONValue
    @State private var text: String?
    var body: some View {
        Group {
            if let text { ToolTextContent(language: "json", code: text) }
            else { ProgressView() }
        }
        .task(id: value) {
            let value = value
            let rendered = await Task.detached(priority: .userInitiated) { chatPrettyJSON(value) }.value
            guard !Task.isCancelled else { return }
            text = rendered
        }
    }
}

/// Legacy asynchronous prompts have a different input shape and no answer map.
private struct QuestionsReadOnlyView: View {
    let input: JSONValue?

    var body: some View {
        let questions = input?[chatKey: "questions"]?.chatArray ?? []
        if questions.isEmpty {
            GenericJSONInput(input: input)
        }
        VStack(alignment: .leading, spacing: 8) {
            ForEach(Array(questions.enumerated()), id: \.offset) { _, entry in
                if let question = entry.chatObject,
                   chatInputString(entry, ["question", "title"]) != nil {
                    VStack(alignment: .leading, spacing: 2) {
                        if let header = question["header"]?.chatString {
                            Text(header)
                                .font(.footnote.weight(.semibold))
                        }
                        if let text = question["question"]?.chatString ?? question["title"]?.chatString {
                            Text(text)
                                .font(.subheadline)
                        }
                        ForEach(Array(optionLabels(question).enumerated()), id: \.offset) { _, label in
                            Text("◦ \(label)")
                                .font(.footnote)
                                .foregroundStyle(.secondary)
                                .padding(.leading, 8)
                                .padding(.top, 2)
                        }
                    }
                } else {
                    GenericJSONInput(input: entry)
                }
            }
        }
    }

    private func optionLabels(_ question: [String: JSONValue]) -> [String] {
        guard let options = question["options"]?.chatArray else { return [] }
        return options.map(toolQuestionOptionText)
    }
}

// MARK: - Result

private struct ToolResultSection: View {
    let tool: ChatToolCall
    @State private var rendering: ResultRendering?
    @State private var prepared = false

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            SectionLabel(text: tool.state == .error
                ? String(localized: "Result · error") : String(localized: "Result"))
            let metadata = toolResultMetadata(tool.result)
            if !metadata.isEmpty {
                Text(metadata.joined(separator: " · "))
                    .font(.caption.monospaced()).foregroundStyle(.secondary)
                    .textSelection(.enabled)
            }
            if let rendering {
                switch rendering {
                case .diffs(let files): DiffTextView(files: files)
                case .terminal(let text): ToolTextContent(language: nil, code: text, terminal: true, isError: tool.state == .error)
                case .json(let text): ToolTextContent(language: "json", code: text)
                case .code(let text, let language): ToolTextContent(language: language, code: text)
                case .markdown(let text): CachedMarkdownView(markdown: text)
                }
            } else if !prepared {
                ProgressView()
            } else {
                Text(tool.state == .running || tool.state == .pending
                     ? String(localized: "Waiting for output…") : String(localized: "No output"))
                    .font(.footnote).foregroundStyle(.secondary)
            }
        }
        .task(id: tool) {
            let tool = tool
            let next = await Task.detached(priority: .userInitiated) { resultRendering(tool) }.value
            guard !Task.isCancelled else { return }
            rendering = next
            prepared = true
        }
    }
}

enum ResultRendering: Sendable {
    case diffs([DiffFile])
    case terminal(String)
    case json(String)
    case code(String, String?)
    case markdown(String)
}

func resultRendering(_ tool: ChatToolCall) -> ResultRendering? {
    guard let result = tool.result, result != .null else { return nil }
    if let text = extractResultText(result) {
        if text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { return nil }
        // Huge diffs use paged source, rather than eagerly making a SwiftUI
        // row for every line. The complete received content stays accessible.
        switch toolResultStyle(tool) {
        case .code(let language): return .code(text, language)
        case .markdown:
            // Do not split markdown mid-fence; large documents use paged source.
            return text.count <= toolTextPageSize ? .markdown(text) : .code(text, "markdown")
        case .terminal:
            if tool.state != .error, text.count <= toolTextPageSize, let files = tryParseDiff(text) { return .diffs(files) }
            return .terminal(text)
        }
    }
    return .json(chatPrettyJSON(result))
}

private struct ToolDiffContent: View {
    let text: String
    @State private var files: [DiffFile]?
    var body: some View {
        Group {
            if let files { DiffTextView(files: files) }
            else { ToolTextContent(language: "diff", code: text) }
        }
        .task(id: text) {
            let text = text
            let parsed = await Task.detached(priority: .userInitiated) {
                text.count <= toolTextPageSize ? tryParseDiff(text) : nil
            }.value
            guard !Task.isCancelled else { return }
            files = parsed
        }
    }
}

// MARK: - Shared helpers

/// Parse `text` as a unified diff when it plausibly is one (the same marker
/// heuristics as the Android port, over the HapiUI parser).
func tryParseDiff(_ text: String) -> [DiffFile]? {
    guard hasDiffMarkers(text) else { return nil }
    let files = UnifiedDiffParser.parse(text)
    guard !files.isEmpty, files.contains(where: { !$0.hunks.isEmpty || $0.isBinary }) else {
        return nil
    }
    return files
}

private func hasDiffMarkers(_ text: String) -> Bool {
    var sawHunk = false
    var sawHeader = false
    for rawLine in text.split(separator: "\n", omittingEmptySubsequences: false) {
        if !sawHunk, rawLine.hasPrefix("@@ -"), rawLine.dropFirst(4).first?.isNumber == true {
            sawHunk = true
        }
        if !sawHeader, rawLine.hasPrefix("diff --git ") || rawLine.hasPrefix("--- ") {
            sawHeader = true
        }
        if sawHunk && sawHeader {
            return true
        }
    }
    return false
}

private let extensionLanguages: [String: String] = [
    "kt": "kotlin", "kts": "kotlin", "java": "java", "ts": "typescript",
    "tsx": "typescript", "js": "javascript", "jsx": "javascript", "py": "python",
    "rb": "ruby", "go": "go", "rs": "rust", "swift": "swift", "c": "c",
    "h": "c", "cpp": "cpp", "cc": "cpp", "cs": "csharp", "sh": "shell",
    "bash": "shell", "json": "json", "yml": "yaml", "yaml": "yaml",
    "cjs": "javascript", "mjs": "javascript", "mts": "typescript", "cts": "typescript",
    "toml": "toml", "zsh": "shell", "diff": "diff", "patch": "diff",
    "xml": "xml", "html": "html", "css": "css", "md": "markdown", "sql": "sql",
]

func languageForPath(_ path: String?) -> String? {
    guard let name = path?.split(whereSeparator: { $0 == "/" || $0 == "\\" }).last?.lowercased() else { return nil }
    if name == "dockerfile" { return "dockerfile" }
    if name == "makefile" { return "makefile" }
    guard let dot = name.lastIndex(of: "."), dot != name.startIndex else { return nil }
    return extensionLanguages[String(name[name.index(after: dot)...])]
}

/// `(glyph, text)` rows for TodoWrite `todos` / update_plan `plan` items.
func checklistItems(_ input: JSONValue?) -> [(glyph: String, text: String)] {
    guard let object = input?.chatObject else { return [] }
    guard let array = (object["todos"] ?? object["plan"])?.chatArray else { return [] }
    return array.compactMap { entry in
        guard let item = entry.chatObject else { return nil }
        guard let content = item["content"]?.chatString ?? item["step"]?.chatString else {
            return nil
        }
        let glyph: String
        switch item["status"]?.chatString {
        case "completed", "complete", "done": glyph = "☑"
        case "in_progress": glyph = "◐"
        default: glyph = "☐"
        }
        return (glyph: glyph, text: content)
    }
}

struct SectionLabel: View {
    let text: String

    var body: some View {
        Text(text)
            .font(.caption2)
            .foregroundStyle(.secondary)
    }
}

File: android/app/src/main/kotlin/app/hapi/companion/feature/chat/blocks/ToolBodies.kt
package app.hapi.companion.feature.chat.blocks

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.produceState
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import app.hapi.companion.R
import app.hapi.companion.feature.chat.terminalCommand
import app.hapi.companion.ui.components.DiffView
import app.hapi.companion.ui.markdown.Markdown
import app.hapi.companion.ui.theme.hapi
import app.hapi.protocol.chat.ChatToolCall
import app.hapi.protocol.chat.getInputString
import app.hapi.protocol.chat.getInputStringAny
import app.hapi.protocol.git.DiffFile
import app.hapi.protocol.git.UnifiedDiffParser
import app.hapi.protocol.wire.HapiJson
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

/**
 * Expanded tool-card body: input rendering per tool kind + the result section
 * (the read-only slice of `web/src/components/ToolCard/views/`):
 *
 * - terminal family → command as a bash code block, stdout/stderr terminal-styled;
 * - `Edit`/`MultiEdit` structured edits → before/after code blocks (the web
 *   derives a word diff from `old_string`/`new_string`; ported minimally);
 * - `Write` → the written content as a code block;
 * - `CodexDiff` (and any input/result that parses as a unified diff) → [DiffView];
 * - `TodoWrite`/`update_plan` → checklist rows;
 * - `ExitPlanMode`/`exit_plan_mode` → complete Markdown proposal from input;
 * - Ask/RequestUserInput → questions + selected answers, read-only;
 * - anything else → pretty-printed JSON input, then the generic result.
 */
@Composable
internal fun ToolCallBody(tool: ChatToolCall, basePath: String?, modifier: Modifier = Modifier) {
    val questionTool = isQuestionDetailsTool(tool.name)
    val plan = planProposalMarkdown(tool)
    val answers = if (questionTool) tool.permission?.answers else null
    Column(modifier = modifier, verticalArrangement = Arrangement.spacedBy(8.dp)) {
        if (questionTool) {
            QuestionToolBody(tool)
        } else if (plan != null) {
            // Plans are reading documents, never paged/truncated tool source.
            Markdown(text = plan)
            if (planProposalShowsResult(tool)) ToolResultSection(tool)
        } else {
            SectionLabel(stringResource(R.string.settings_usage_input))
            ToolInputSection(tool)
            ToolResultSection(tool)
        }
        var sourceExpanded by rememberSaveable(tool.id) { mutableStateOf(false) }
        if (tool.input != null || tool.result != null || answers != null) {
            TextButton(onClick = { sourceExpanded = !sourceExpanded }) {
                Text(stringResource(R.string.files_viewer_source))
            }
            if (sourceExpanded) {
                tool.input?.let {
                    SectionLabel(stringResource(R.string.settings_usage_input))
                    GenericJsonInput(it)
                }
                tool.result?.let {
                    SectionLabel(stringResource(R.string.chat_result))
                    GenericJsonInput(it)
                }
                answers?.let {
                    SectionLabel(stringResource(R.string.tool_question_answers))
                    GenericJsonInput(it)
                }
            }
        }
    }
}

@Composable
private fun QuestionToolBody(tool: ChatToolCall) {
    val details by produceState<QuestionToolDetails?>(null, tool) {
        value = withContext(Dispatchers.Default) { questionToolDetails(tool) }
    }
    val prepared = details
    if (prepared == null) {
        androidx.compose.material3.CircularProgressIndicator()
    } else {
        SectionLabel(stringResource(if (prepared.hasAnswers) R.string.tool_questions_answers else R.string.settings_usage_input))
        if (prepared.questions.isEmpty()) GenericJsonInput(tool.input)
        else QuestionDetailsView(prepared.questions)
        if (prepared.showResult) ToolResultSection(tool)
    }
}

// ------------------------------------------------------------------ input --

@Composable
private fun ToolInputSection(tool: ChatToolCall) {
    val input = tool.input
    val name = toolPresentationName(tool.name)
    when {
        name in TERMINAL_TOOLS -> {
            val command = terminalCommand(input)
            if (command != null) ToolTextContent(code = command, language = "bash") else GenericJsonInput(input)
        }

        name == "exec" -> {
            val source = toolSourceInput(input, listOf("code", "script"))
            if (source != null) ToolTextContent(code = source, language = "javascript") else GenericJsonInput(input)
        }

        name == "CodexPatch" -> {
            val patch = toolSourceInput(input, listOf("patch", "input", "command"))
            if (patch != null) ToolTextContent(code = patch, language = "diff") else GenericJsonInput(input)
        }

        name == "Edit" -> {
            val old = getInputString(input, "old_string")
            val new = getInputString(input, "new_string")
            if (old != null && new != null) {
                BeforeAfter(old, new, languageForPath(getInputStringAny(input, listOf("file_path", "path"))))
            } else {
                GenericJsonInput(input)
            }
        }

        name == "MultiEdit" -> {
            val language = languageForPath(getInputStringAny(input, listOf("file_path", "path")))
            val edits = (input as? JsonObject)?.get("edits") as? JsonArray
            if (!edits.isNullOrEmpty()) {
                edits.forEachIndexed { index, edit ->
                    val old = getInputString(edit, "old_string")
                    val new = getInputString(edit, "new_string")
                    if (old != null && new != null) {
                        if (edits.size > 1) {
                            SectionLabel(stringResource(R.string.chat_edit_n_of_m, index + 1, edits.size))
                        }
                        BeforeAfter(old, new, language)
                    } else {
                        GenericJsonInput(edit)
                    }
                }
            } else {
                GenericJsonInput(input)
            }
        }

        name == "Write" -> {
            val content = getInputStringAny(input, listOf("content", "text"))
            if (content != null) {
                ToolTextContent(
                    code = content,
                    language = languageForPath(getInputStringAny(input, listOf("file_path", "path"))),
                )
            } else {
                GenericJsonInput(input)
            }
        }

        name == "CodexDiff" -> {
            val unified = getInputString(input, "unified_diff")
            val files = remember(unified) { unified?.takeIf { fitsToolPage(it) }?.let(::tryParseDiff) }
            if (files != null) {
                files.forEach { DiffView(file = it) }
            } else if (unified != null) {
                ToolTextContent(code = unified, language = "diff")
            } else {
                GenericJsonInput(input)
            }
        }

        name == "TodoWrite" || name == "update_plan" -> {
            val items = checklistItems(input)
            if (items.isNotEmpty()) {
                Column(verticalArrangement = Arrangement.spacedBy(2.dp)) {
                    items.forEach { (state, text) ->
                        Text(
                            text = "$state $text",
                            style = MaterialTheme.typography.bodySmall,
                        )
                    }
                }
            } else {
                GenericJsonInput(input)
            }
        }

        name == "request_user_input_async" -> {
            QuestionsReadOnly(input)
        }

        else -> GenericJsonInput(input)
    }
}

@Composable
private fun BeforeAfter(old: String, new: String, language: String?) {
    Column(verticalArrangement = Arrangement.spacedBy(6.dp)) {
        val emptyLabel = stringResource(R.string.chat_empty_snippet)
        SectionLabel(stringResource(R.string.chat_before))
        ToolTextContent(code = old.ifEmpty { emptyLabel }, language = language)
        SectionLabel(stringResource(R.string.chat_after))
        ToolTextContent(code = new.ifEmpty { emptyLabel }, language = language)
    }
}

@Composable
private fun GenericJsonInput(input: JsonElement?) {
    when {
        input == null || input is JsonNull -> Unit
        input is JsonPrimitive && input.isString -> ToolTextContent(code = input.content, language = null)
        else -> {
            val text by produceState<String?>(null, input) {
                value = null
                value = withContext(Dispatchers.Default) { prettyJson(input) }
            }
            text?.let { ToolTextContent(code = it, language = "json") }
        }
    }
}

@Composable
private fun QuestionsReadOnly(input: JsonElement?) {
    val questions = (input as? JsonObject)?.get("questions") as? JsonArray
    if (questions.isNullOrEmpty()) {
        GenericJsonInput(input)
        return
    }
    val hint = MaterialTheme.hapi.hint
    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
        questions.forEach { entry ->
            val question = entry as? JsonObject
            if (question == null || getInputStringAny(question, listOf("question", "title")) == null) {
                GenericJsonInput(entry)
                return@forEach
            }
            val header = (question["header"] as? JsonPrimitive)?.contentOrNullIfNotString()
            val text = getInputStringAny(question, listOf("question", "title"))
            Column {
                header?.let {
                    Text(text = it, style = MaterialTheme.typography.labelLarge, fontWeight = FontWeight.SemiBold)
                }
                text?.let {
                    Text(text = it, style = MaterialTheme.typography.bodyMedium)
                }
                val options = question["options"] as? JsonArray
                options?.forEach { option ->
                    Text(
                        text = "◦ ${toolQuestionOptionText(option)}",
                        style = MaterialTheme.typography.bodySmall,
                        color = hint,
                        modifier = Modifier.padding(start = 8.dp, top = 2.dp),
                    )
                }
            }
        }
    }
}

// ----------------------------------------------------------------- result --

/** How a tool result renders: parsed diff > extracted text > pretty JSON. */
internal sealed interface ResultRendering {
    data class Diffs(val files: List<DiffFile>) : ResultRendering
    data class Terminal(val text: String) : ResultRendering
    data class Json(val pretty: String) : ResultRendering
    data class Code(val text: String, val language: String?) : ResultRendering
    data class Prose(val text: String) : ResultRendering
}

@Composable
private fun ToolResultSection(tool: ChatToolCall) {
    val result = tool.result ?: return
    if (result is JsonNull) return
    val isError = tool.state == "error"
    val rendering by produceState<ResultRendering?>(null, tool) {
        value = withContext(Dispatchers.Default) { resultRendering(tool) }
    }

    Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
        SectionLabel(
            stringResource(if (isError) R.string.chat_result_error else R.string.chat_result),
        )
        val metadata = remember(result) { toolResultMetadata(result) }
        if (metadata.isNotEmpty()) SectionLabel(metadata.joinToString(" · "))
        when (val rendering = rendering) {
            is ResultRendering.Diffs -> rendering.files.forEach { DiffView(file = it) }
            is ResultRendering.Terminal -> ToolTextContent(code = rendering.text, terminal = true, isError = isError)
            is ResultRendering.Json -> ToolTextContent(code = rendering.pretty, language = "json")
            is ResultRendering.Code -> ToolTextContent(code = rendering.text, language = rendering.language)
            is ResultRendering.Prose -> Markdown(text = rendering.text)
            null -> Unit
        }
    }
}

internal fun resultRendering(tool: ChatToolCall): ResultRendering? {
    val result = tool.result ?: return null
    if (result is JsonNull) return null
    val text = extractResultText(result)
    if (text != null) {
        if (text.isBlank()) return null
        return when (val style = toolResultStyle(tool)) {
            is ToolResultStyle.Code -> ResultRendering.Code(text, style.language)
            ToolResultStyle.Markdown -> if (fitsToolPage(text)) ResultRendering.Prose(text)
                else ResultRendering.Code(text, "markdown")
            ToolResultStyle.Terminal -> {
                if (tool.state != "error" && fitsToolPage(text)) {
                    tryParseDiff(text)?.let { return ResultRendering.Diffs(it) }
                }
                ResultRendering.Terminal(text)
            }
        }
    }
    return ResultRendering.Json(prettyJson(result))
}

// ---------------------------------------------------------------- helpers --

private val DIFF_MARKER = Regex("(^|\n)@@ -\\d")
private val DIFF_HEADER = Regex("(^|\n)(diff --git |--- )")

/** Parse [text] as a unified diff when it plausibly is one. */
internal fun tryParseDiff(text: String): List<DiffFile>? {
    if (!DIFF_MARKER.containsMatchIn(text) || !DIFF_HEADER.containsMatchIn(text)) return null
    val files = UnifiedDiffParser.parse(text)
    return files.takeIf { parsed -> parsed.isNotEmpty() && parsed.any { it.hunks.isNotEmpty() || it.isBinary } }
}

private val prettyJsonFormat = Json(from = HapiJson) { prettyPrint = true }

internal fun prettyJson(element: JsonElement): String =
    prettyJsonFormat.encodeToString(JsonElement.serializer(), element)

private val EXTENSION_LANGUAGES = mapOf(
    "kt" to "kotlin", "kts" to "kotlin", "java" to "java", "ts" to "typescript",
    "tsx" to "typescript", "js" to "javascript", "jsx" to "javascript", "py" to "python",
    "rb" to "ruby", "go" to "go", "rs" to "rust", "swift" to "swift", "c" to "c",
    "h" to "c", "cpp" to "cpp", "cc" to "cpp", "cs" to "csharp", "sh" to "shell",
    "bash" to "shell", "json" to "json", "yml" to "yaml", "yaml" to "yaml",
    "cjs" to "javascript", "mjs" to "javascript", "mts" to "typescript", "cts" to "typescript",
    "toml" to "toml", "zsh" to "shell", "diff" to "diff", "patch" to "diff",
    "xml" to "xml", "html" to "html", "css" to "css", "md" to "markdown", "sql" to "sql",
)

internal fun languageForPath(path: String?): String? {
    val name = path?.replace('\\', '/')?.substringAfterLast('/')?.lowercase() ?: return null
    if (name == "dockerfile") return "dockerfile"
    if (name == "makefile") return "makefile"
    if (name.startsWith('.') && name.count { it == '.' } == 1) return null
    return EXTENSION_LANGUAGES[name.substringAfterLast('.', missingDelimiterValue = "")]
}

private fun JsonPrimitive.contentOrNullIfNotString(): String? = if (isString) content else null

/** `(glyph, text)` rows for TodoWrite `todos` / update_plan `plan` items. */
private fun checklistItems(input: JsonElement?): List<Pair<String, String>> {
    val obj = input as? JsonObject ?: return emptyList()
    val array = (obj["todos"] as? JsonArray) ?: (obj["plan"] as? JsonArray) ?: return emptyList()
    return array.mapNotNull { entry ->
        val item = entry as? JsonObject ?: return@mapNotNull null
        val content = (item["content"] as? JsonPrimitive)?.contentOrNullIfNotString()
            ?: (item["step"] as? JsonPrimitive)?.contentOrNullIfNotString()
            ?: return@mapNotNull null
        val status = (item["status"] as? JsonPrimitive)?.contentOrNullIfNotString()
        val glyph = when (status) {
            "completed", "complete", "done" -> "☑"
            "in_progress" -> "◐"
            else -> "☐"
        }
        glyph to content
    }
}

@Composable
private fun SectionLabel(text: String) {
    Text(
        text = text,
        style = MaterialTheme.typography.labelSmall,
        color = MaterialTheme.hapi.hint,
    )
}
